packages/roosterjs-editor-plugins/lib/plugins/ContentEdit/features/markdownFeatures.ts

```diff
@@ -0,0 +1,127 @@
+import { ChangeSource, NodePosition, PluginKeyboardEvent } from 'roosterjs-editor-types';
+import { ContentEditFeature, Editor, Keys } from 'roosterjs-editor-core';
+import { createRange } from 'roosterjs-editor-dom';
+
+const ZERO_WIDTH_SPACE = '\u200B';
+
+function generateBasicMarkdownFeature(
+    key: Keys,
+    triggerCharacter: string,
+    elementTag: string,
+    useShiftKey: boolean
+): ContentEditFeature {
+    return {
+        keys: [key],
+        shouldHandleEvent: (event, editor) =>
+            event.rawEvent.shiftKey === useShiftKey &&
+            !!getRangeForMarkdownOperation(event, editor, triggerCharacter),
+        handleEvent: (event, editor) => {
+            event.rawEvent.preventDefault();
+            handleMarkdownEvent(event, editor, triggerCharacter, elementTag);
+        },
+    };
+}
+
+function getRangeForMarkdownOperation(
+    event: PluginKeyboardEvent,
+    editor: Editor,
+    triggerCharacter: string
+): Range {
+    const searcher = editor.getContentSearcherOfCursor();
+
+    let startPosition: NodePosition;
+    let endPosition: NodePosition;
+    searcher.forEachTextInlineElement(textInlineElement => {
+        if (endPosition && startPosition) {
+            return true;
+        }
+        const inlineTextContent = textInlineElement.getTextContent();
+        if (inlineTextContent[inlineTextContent.length - 1].trim().length == 0) {
+            return false;
+        }
+        endPosition = textInlineElement.getStartPosition().move(inlineTextContent.length);
+        if (inlineTextContent[0] == triggerCharacter) {
+            startPosition = textInlineElement.getStartPosition();
+        } else {
+            let contentIndex = inlineTextContent.length - 1;
+            for (; contentIndex > 0; contentIndex--) {
+                if (startPosition) {
+                    return true;
+                }
+                if (
+                    inlineTextContent[contentIndex] == triggerCharacter &&
+                    inlineTextContent[contentIndex - 1].trim().length == 0
+                ) {
+                    startPosition = textInlineElement.getStartPosition().move(contentIndex);
+                    return true;
+                }
+            }
+        }
+    });
+    return !!startPosition && !!endPosition && createRange(startPosition, endPosition);
+}
+
+function handleMarkdownEvent(
+    event: PluginKeyboardEvent,
+    editor: Editor,
+    triggerCharacter: string,
+    elementTag: string
+) {
+    editor.performAutoComplete(() => {
```

Comment:
editor.runAsync() might be necessary here for now, otherwise the "backspace to undo auto complete" feature won't work correctly because the snapshot before auto complete will be immediately after it is taken.